src/librustc/middle/ty.rs

```diff
@@ -6100,13 +6100,7 @@ impl<'tcx> ctxt<'tcx> {
                 let found = match val {
                     ConstVal::Uint(count) => return count as usize,
                     ConstVal::Int(count) if count >= 0 => return count as usize,
-                    ConstVal::Int(_) => "negative integer",
-                    ConstVal::Float(_) => "float",
-                    ConstVal::Str(_) => "string",
-                    ConstVal::Bool(_) => "boolean",
-                    ConstVal::Binary(_) => "binary array",
-                    ConstVal::Struct(..) => "struct",
-                    ConstVal::Tuple(_) => "tuple"
+                    const_val => const_val.description(),
```

Comment:
You might want to keep around the special-case for ConstVal::Int?  The fact that the integer is negative is important.